services/m3aggregator/config/aggregator.go

```diff
@@ -531,50 +551,72 @@ func (c *kvClientConfiguration) NewKVClient(instrumentOpts instrument.Options) (
 	return c.Etcd.NewClient(instrumentOpts)
 }
 
-type shardFnType string
+type hashFnType string
 
-// List of supported sharding function types.
+// List of supported hashing function types.
 const (
-	murmur32ShardFn shardFnType = "murmur32"
+	murmur32HashFn hashFnType = "murmur32"
 
-	defaultShardFn = murmur32ShardFn
+	defaultHashFn = murmur32HashFn
 )
 
 var (
-	validShardFnTypes = []shardFnType{
-		murmur32ShardFn,
+	validHashFnTypes = []hashFnType{
+		murmur32HashFn,
 	}
 )
 
-func (t *shardFnType) UnmarshalYAML(unmarshal func(interface{}) error) error {
+func (t *hashFnType) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	var str string
 	if err := unmarshal(&str); err != nil {
 		return err
 	}
 	if str == "" {
-		*t = defaultShardFn
+		*t = defaultHashFn
 		return nil
 	}
-	validTypes := make([]string, 0, len(validShardFnTypes))
-	for _, valid := range validShardFnTypes {
+	validTypes := make([]string, 0, len(validHashFnTypes))
+	for _, valid := range validHashFnTypes {
 		if str == string(valid) {
 			*t = valid
 			return nil
 		}
 		validTypes = append(validTypes, string(valid))
 	}
-	return fmt.Errorf("invalid shrading function type '%s' valid types are: %s",
+	return fmt.Errorf("invalid hashing function type '%s' valid types are: %s",
 		str, strings.Join(validTypes, ", "))
 }
 
-func (t shardFnType) ShardFn() (aggregator.ShardFn, error) {
+func (t hashFnType) ShardFn() (aggregator.ShardFn, error) {
 	switch t {
-	case murmur32ShardFn:
+	case murmur32HashFn:
 		return func(id []byte, numShards int) uint32 {
 			return murmur3.Sum32(id) % uint32(numShards)
 		}, nil
 	default:
-		return nil, fmt.Errorf("unrecognized hash gen type %v", t)
+		return nil, fmt.Errorf("unrecognized shard function type %v", t)
+	}
+}
+
+func (t hashFnType) PartitionFnGen(totalPartitions int) (aggregator.PartitionFnGen, error) {
+	switch t {
+	case murmur32HashFn:
+		// NB(xichen): If this turns out to be too CPU intensive due to byte copies, can rewrite
+		// it to compute murmur3 hashes with zero byte copies.
+		return func() aggregator.PartitionFn {
+			// Create a new local buffer on every function invocation, therefore
+			// making the generated partition functions independent of each other.
+			buf := bytes.NewBuffer(make([]byte, 0, initialChunkedIDSize))
```

Comment:
literally just read a blog post that might be relevant here: https://segment.com/blog/allocation-efficiency-in-high-performance-go-services/

I haven't done any benchmarks but I think it might be more efficient to do the following:

```
return func(chunkedID id.ChunkedID) uint32 {
    var b [256]byte
    buf := b[:0]

    buf = append(buf, chunkedID.Prefix...)
    buf = append(buf, chunkedID.Data...)
    buf = append(buf, chunkedID.Suffix...)
    return murmur3.Sum32(buf) % uint32(totalPartitions)
}
```

This *should* only allocate when the total ID length is greater than 256. Need to run the benchmarks to confirm.

Reply:
Huzzah! It worked!

Benchmark:

```
package main

import (
	"bytes"
	"testing"

	"github.com/spaolacci/murmur3"
)

var (
	foo = []byte("foo")
	bar = []byte("bar")
	baz = []byte("baz")
)

func BenchmarkBuffer(b *testing.B) {
	var hash uint32
	for n := 0; n < b.N; n++ {
		buf := bytes.NewBuffer(make([]byte, 0, 16))
		buf.Write(foo)
		buf.Write(bar)
		buf.Write(baz)
		hash = murmur3.Sum32(buf.Bytes())
	}
	_ = hash
}

func BenchmarkArray(b *testing.B) {
	var hash uint32
	for n := 0; n < b.N; n++ {
		var a [16]byte
		buf := a[:0]
		buf = append(buf, foo...)
		buf = append(buf, bar...)
		buf = append(buf, baz...)
		hash = murmur3.Sum32(buf)
	}
	_ = hash
}
```

Result:

```
BenchmarkBuffer-4       10000000               123 ns/op             128 B/op          2 allocs/op
BenchmarkArray-4        50000000                24.9 ns/op             0 B/op          0 allocs/op
```